### PR TITLE
SIMD-0388: BLS12-381 Elliptic Curve Syscalls

### PR DESCRIPTION
**Pull request by samkim-crypto**
Add BLS12-381 elliptic curve syscalls

### CODE REVIEW

File: proposals/0388-bls12-381-syscalls.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+foundation for many efficient and modern zero-knowledge proof systems, such as
+those based on Groth16. While the BN254 curve enables basic cryptography and
+zero-knowledge applications, it does not provide a 128-bit security level, which
+is insufficient for high-security protocols.
+
+The BLS12-381 curve is a modern, widely adopted standard for a pairing-friendly
+curve that achieves a 128-bit security level. It is adopted in other major
+ecosystems, such as Ethereum, which already supports it via a precompile.
+
+Adding support for the BLS12-381 curve is important for enabling Alpenglow consensus
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md:39]: nit: we could link to SIMD 326 here.

File: proposals/0388-bls12-381-syscalls.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+as well. Alpenglow intends to process standard vote transactions by verifying
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: nit: process consensus votes, we don't have concept of vote transactions any more.

File: proposals/0388-bls12-381-syscalls.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+BLS signatures in a dedicated BLS sigverify stage within the validator logic, so
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: hmm, I think whether it's dedicated BLS sigverify stage is internal implementation, just say BLS verification on votes doesn't need the syscall, we are calling the library directly.

File: proposals/0388-bls12-381-syscalls.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+a BLS12-381 syscall is not necessary for the vote transactions themselves.
+
+However, when a validator registers their BLS public key, it must do so using a
+regular on-chain transaction that will pass through the normal transaction
+pipeline. This transaction must contain a BLS Proof of Possession (PoP) to
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: nit: Do you think we should maybe list "BLS Proof of Possession" in New Terminology section? This is new to the chain.

File: proposals/0388-bls12-381-syscalls.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+validate ownship of the BLS public key. This PoP is a cryptographic proof that
+is crucial to prevent a "rogue-key-attack", where a malicious validator could,
+roughly speaking, try to register another validator's public key as their own.
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: nit: we can remove "roughly speaking" maybe?

File: proposals/0388-bls12-381-syscalls.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+
+1. The addition, subtraction (negation) and scalar multiplication in G1
+2. The pairing operation
+3. The decompression operations in G1 and G2
+
+These operations are sufficient to enable standard pairing-based cryptography
+applications, for example, Groth16 proof verifier on BLS12-381 and BLS signature
+or Proof Of Possession (PoP) verifier. Additional operations are needed to
+support more advanced ZK or cryptography applications, but these are outside the
+scope of this SIMD for now.
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: nit: remove "for now".

File: proposals/0388-bls12-381-syscalls.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+elliptic curve group operations. This function takes in a `curve_id`, `group_op`,
+and two scalar/points that are encoded in little-endian. It interprets the
+curve points according to the curve id, and applies the group operations
+specified by the `group_op`. Currently, the syscall supports the curve25519
+edwards and ristretto representations.
+
+This function can be extended to support the addition, subtraction, and scalar
+multiplication in BLS12-381 G1. We propose adding new `curve_id` constants for
+BLS12-381. The BLS12-381 inputs should be interpreted as affine points unlike
+how Curve25519 points are currently interpreted by the syscall.
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: nit: they use inputs differently? Maybe we just list which is used by BLS12-381 and which is used by Curve25519?

File: proposals/0388-bls12-381-syscalls.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+define_syscall!(fn sol_curve_pairing_map(
+    curve_id: u64,
+    point: *const u8,
+    result: *mut u8
+) -> u64);
+```
+
+This function will take in a curve id and a point that is represented in its
+compressed representation (encoded in little-endian). It will decompress the
+compressed point into an affine representation and write the result.
```
Comment (wen-coding) [inline proposals/0388-bls12-381-syscalls.md]: Hmm, do you mean, if I have a compressed pubkey and compressed signature, I need to do three syscalls to verify proof of possession? Would that be more expensive than one syscall?

File: proposals/0388-bls12-381-syscalls.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+encoded in little-endian. It should then compute the pairing product
+`e(P1, Q1) * ... * e(Pn, Qn)`.
+
+### Decompression Operations in G1 and G2
+
+For the decompression operations in G1 and G2, we propose adding a dedicated
+syscall for general decompression `sol_curve_decompress`.
+
+```rust
+define_syscall!(fn sol_curve_pairing_map(
```
Comment (0x0ece) [inline proposals/0388-bls12-381-syscalls.md]: nit: sol_curve_decompress

File: proposals/0388-bls12-381-syscalls.md (hunk 8 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md]: Right, the PoP will actually use 4 syscalls:
1. sha256 to hash the public key
2. decompress public key
3. decompress signature (PoP)
4. pairing

There is some additional overhead for invoking a syscall itself. However, the cryptographic operations themselves are relatively much more costly, so the actual overhead of invoking multiple syscall as opposed to executing each of these in one larger syscall should be almost negligible I believe. Unlike votes, PoP is done once for each validator during BLS public key registration, so I think the cost should be reasonable.

File: proposals/0388-bls12-381-syscalls.md [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+This function is not actually instantiated at the moment. We propose updating
+this function's signature to take in an array of points in G1 and an array of
+pointst in G2 to support batch pairings.
+
+```rust
+define_syscall!(fn sol_curve_pairing_map(
+    curve_id: u64,
+    g1_points: *const u8,
+    g2_points: *const u8,
+    result: *mut u8
```
Comment (ananas-block) [inline proposals/0388-bls12-381-syscalls.md:254]: Is the intention for the pairing result to be a `bool` (bn254 syscall) or the actual result?

Returning the actual result would reduce the number of pairs necessary to verify a Groth16 proof by one from 4 to 3 (12k CU for the current bn254 syscall).

File: proposals/0388-bls12-381-syscalls.md [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+## Summary
+
+This proposal introduces a new family of syscalls to provide native support for
+cryptographic operations on the BLS12-381 elliptic curve. These syscalls will
+expose:
+
+1. The addition, subtraction (negation) and scalar multiplication in G1
+2. The pairing operation
+3. The decompression operations in G1 and G2
+
```
Comment (deanmlittle) [inline proposals/0388-bls12-381-syscalls.md:23]: We should also add G2 addition and multiplication. It unlocks Halo2 proofs, hash to G2 onchain, and verification of external Alpenglow quorums so we can actually have real, L1-verified L2s on mainnet.

File: proposals/0388-bls12-381-syscalls.md [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+// Reserve indices 2 and 3 in case we want to support affine representations of
+// curve25519 points in the future
+// pub const CURVE25519_EDWARDS_AFFINE_LE: u64 = 2;
+// pub const CURVE25519_EDWARDS_AFFINE_BE: u64 = 2 | 0x80;
+// pub const CURVE25519_RISTRETTO_AFFINE_LE: u64 = 3;
+// pub const CURVE25519_RISTRETTO_AFFINE_BE: u64 = 3 | 0x80;
+
+// New Curve ID
+pub const BLS12_381_LE: u64 = 4;
+pub const BLS12_381_BE: u64 = 4 | 0x80;
```
Comment (deanmlittle) [inline proposals/0388-bls12-381-syscalls.md:146]: Is there any demand for BE? 🤔

File: proposals/0388-bls12-381-syscalls.md [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+([SIMD 326](https://github.com/solana-foundation/solana-improvement-documents/pull/326))
+as well. Alpenglow consensus votes themselves do not require a BLS12-381
+syscall as they are processed differently than regular transactions.
+
+However, when a validator registers their BLS public key, it must do
+so using a regular on-chain transaction that will pass through the normal
+transaction pipeline. This transaction must contain a BLS Proof of Possession
+(PoP) to validate ownership of the BLS public key. This PoP is a cryptographic
+proof that is crucial to prevent a "rogue-key-attack", where a malicious
+validator could try to register a a key that is an algebraic combination of
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md]:
```suggestion
validator could try to register a key that is an algebraic combination of
```

File: proposals/0388-bls12-381-syscalls.md [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    result: *mut u8
+) -> u64);
+```
+
+This function is not actually instantiated at the moment. We propose updating
+this function's signature to take in an array of points in G1 and an array of
+pointst in G2 to support batch pairings.
+
+```rust
+define_syscall!(fn sol_curve_pairing_map(
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md:249]: With this syscall prototype, how will you know how many pairs are provided in the arrays? `sol_alt_bn128_group_op` takes in a length parameter.

File: proposals/0388-bls12-381-syscalls.md [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+    curve_id: u64,
+    point: *const u8,
+    result: *mut u8
+) -> u64);
+```
+
+This function is not actually instantiated at the moment. We propose updating
+this function's signature to take in an array of points in G1 and an array of
+pointst in G2 to support batch pairings.
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md]:
```suggestion
points in G2 to support batch pairings.
```

File: proposals/0388-bls12-381-syscalls.md [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+) -> u64);
+```
+
+The function would interpret `g1_points` as an array of BLS12-381 points in G1
+`[P1, ..., Pn]` and `g2_points` as an array of BLS12-381 points in G2
+`[Q1, ..., Qn]`. Both inputs are interpreted as affine representations
+encoded in either little-endian or big-endian depending on the curve id. It
+should then compute the pairing product `e(P1, Q1) * ... * e(Pn, Qn)`.
+
+### Decompression Operations in G1 and G2
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md:283]: Would it be possible to clarify the decompression validation behaviour here? Or at least link to where it might be explained. 
What qualifies a valid encoding?
Are we following [the IETF draft?](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html)?

File: proposals/0388-bls12-381-syscalls.md [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+
+// Reserve indices 2 and 3 in case we want to support affine representations of
+// curve25519 points in the future
+// pub const CURVE25519_EDWARDS_AFFINE_LE: u64 = 2;
+// pub const CURVE25519_EDWARDS_AFFINE_BE: u64 = 2 | 0x80;
+// pub const CURVE25519_RISTRETTO_AFFINE_LE: u64 = 3;
+// pub const CURVE25519_RISTRETTO_AFFINE_BE: u64 = 3 | 0x80;
+
+// New Curve ID
+pub const BLS12_381_LE: u64 = 4;
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md:145]: Apologies, this question may have already been answered, but what is the difference between `BLS12_381_LE` and `BLS12_381_G1_LE`?

File: proposals/0388-bls12-381-syscalls.md [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+
+This proposal introduces a new family of syscalls to provide native support for
+cryptographic operations on the BLS12-381 elliptic curve. These syscalls will
+expose:
+
+1. The addition, subtraction (negation) and scalar multiplication in G1
+2. The pairing operation
+3. The decompression operations in G1 and G2
+
+Group operations on G2 and Poseidon hash support is outside the scope of this
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md]: By "Poseidon hash support", you mean adding a syscall which is poseidon hash over bls12-381 instead of bn254, right? Just want to make sure I understand. Not a problem with the SIMD.

File: proposals/0388-bls12-381-syscalls.md (hunk 11 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:23]: Yes, I am in support to add G2 group ops as well eventually, but I wanted to limit the scope of the current proposal since I don't want this to be a blocker for alpenglow (we need the syscalls to verify PoP of BLS public keys).

We might need more development time to add support for G2 (esp. with fd) and maybe we can run into issues like G2 ops being too expensive and what not, which could delay feature gate activation, which can affect alpenglow. I don't think it hurts to separate out G2 ops in a separate SIMD and therefore a separate feature gate.

File: proposals/0388-bls12-381-syscalls.md (hunk 12 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:146]: At least the [rfc](https://www.ietf.org/archive/id/draft-irtf-cfrg-bls-signature-05.html#name-bls12-381) for bls signatures cite the [zcash](https://github.com/zkcrypto/pairing/blob/34aa52b0f7bef705917252ea63e5a13fa01af551/src/bls12_381/README.md#serialization) serialization, which encodes things in big-endian. The `blst` library also serializes using big-endian, so the current bls sig implementation uses BE as well.

Comment (deanmlittle) [reply proposals/0388-bls12-381-syscalls.md:146]: wasn't aware of that. strange choice, but fair enough!

File: proposals/0388-bls12-381-syscalls.md (hunk 10 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:254]: Yeah I was thinking of actually returning the result in the target group, but I realized that I didn't make that clear. I added a clarifying statement.

File: proposals/0388-bls12-381-syscalls.md (hunk 14 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:249]: Oh yes this is a nice catch! We do need to add specify the lengths for each of the inputs. I updated to add the lengths.

File: proposals/0388-bls12-381-syscalls.md (hunk 16 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:283]: This is a good point that I don't fully specify it in the doc. This should follow the Zcash specification. I updated the doc with the specification.

File: proposals/0388-bls12-381-syscalls.md (hunk 17 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:145]: For something like the pairing function, we just need to specify the curve and hence `BLS12_381_{LE,BE}`. But for functions like group op and decompress, we need to specify either G1 or G2 and hence we have `BLS12_381_G1_{LE,BE}` and `BLS12_381_G2_{LE,BE}`. I discuss it below where the constants are specified.

File: proposals/0388-bls12-381-syscalls.md (hunk 18 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md]: Right. I mean extending the poseidon syscall to work over bls12-381 in addition to bn254 (maybe by extending [this](https://github.com/anza-xyz/agave/blob/master/poseidon/src/lib.rs#L118)).

File: proposals/0388-bls12-381-syscalls.md (hunk 14 above)
Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md:249]: What point is there in having them as separate parameters, how could they ever be different lengths?

I do like the SoA approach, rather than AoS approach taken by the alt-bn128 syscall, but I don't see a reason to have two length parameters when the implementation would just end up comparing them, and erroring anyways.

File: proposals/0388-bls12-381-syscalls.md (hunk 16 above)
Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md:283]: This is much better, thank you.

File: proposals/0388-bls12-381-syscalls.md (hunk 18 above)
Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md]: Nice, makes sense.

File: proposals/0388-bls12-381-syscalls.md (hunk 17 above)
Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md:145]: Hm, makes sense. I could see ways to simplify it to just share them, but there is no need for that I imagine, it's all so trivial.

File: proposals/0388-bls12-381-syscalls.md (hunk 10 above)
Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md:254]: I have wondered about this. As I understand it, BLS signature verification and such only cares about 

```math
e(p_1, q_1) \cdot e(p_2, q_2)^{-1} = 1
```

So moving the identity check out into the ebpf makes that usecase slower. I agree there should be a way to get the result of the miller eval out, esp for circuit constraints and stuff, but there are usecases for the identity case.

Perhaps we could specify which happens based on the curve id? I wouldn't want it as another argument, since that just increases register pressure for no real reason.

File: proposals/0388-bls12-381-syscalls.md (hunk 14 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:249]: Yeah I think the alt-bn128 syscall pushes a lot of expensive work onto the BPF program. In Groth16 verifier, for example, the G1 points (from the proof) and G2 poinits (from the verifying keys) generally live in different memory and are almost never stored together in the interleaved format.

In terms of one length parameter vs two, I think having a byte length parameter for each array parameter to a function is just a direct and more canonical C-style API. An alternative would be to pass a `num_pairs` parameter and I think this is fine as well, but the runtime would then need to calculate `expected_g1_len = num_pairs * 96` and `expected_g2_len = num_pairs * 192` and use this number to parse the arrays. I think just having the byte lengths directly follow the arrays feels more natural to me, but I am open to taking in `num_pairs` if others feel strongly about this.

File: proposals/0388-bls12-381-syscalls.md (hunk 10 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:254]: This is a reasonable point. But for the identity check operation, what we need is essentially just a `memcmp`, right? After the result is returned, we just need to compare the result to a hard-coded identity element in the target group, which I can't imagine it being too expensive. For applications like groth16, pinnocchio, groth-maller, etc. we would be saving an entire pairing which is in the realm of tens of thousands of CUs (12k for alt_bn128).

Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md:254]: True, it would just be a 576 byte memcmp (12 * 48), which is 100 CUs (`max(100, 576 / 250)`). So I guess it doesn't matter too much. The reason I brought this up is because of how trivial it is to support.

File: proposals/0388-bls12-381-syscalls.md (hunk 14 above)
Comment (Rexicon226) [reply proposals/0388-bls12-381-syscalls.md:249]: I personally like the `num_pairs` approach myself. If you pass in byte lengths, the runtime still needs to check
```
g1_bytes % (2 * 48) = 0
g2_bytes % (4 * 48) = 0
if (g1_bytes / (2 * 48) != g2_bytes / (4 * 48)) return error
```
which leads to yet another thing the caller can get wrong.
I feel that in most callsites of this syscall, the program will already have the number of pairs. 
Perhaps someone else could chime in and clarify which is more ergonomic for the users?

As for it being more C-style, eh, maybe. APIs like `memcmp` and `memmove` already take in one length parameter which is for two parameters. I think this is comparable.

File: proposals/0388-bls12-381-syscalls.md (hunk 11 above)
Comment (mertimus) [reply proposals/0388-bls12-381-syscalls.md:23]: I don't quite get the risk here, what do we mean by too expensive and what not? blocks are relatively empty and it's a safe change that can enable a ton of ecosystem teams on top to build new things that they need to compromise on today

Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:23]: I will actually go ahead and add G2 group ops in this SIMD since there is a lot of demand for this. 

I was initially hoping to add G2 group ops in a separate SIMD/feature gate to not block alpenglow. Adding G2 would obviously require more development time to test, assign the appropriate CUs, make sure that all the edge cases are handled correctly among validator clients, etc. and I wanted to minimize the chance of alpenglow launch being blocked by the activation of BLS12-381 syscall.

But the most recent alpenglow proposal was [updated](https://github.com/solana-foundation/solana-improvement-documents/pull/387/commits/a8e44f94212516ae791818591bc87cedf53d67eb) to not rely on the bls12-381 syscall, but to perform the BLS PoP verification inside the vote program at launch. So I think we can now have a more complete BLS12-381 syscall proposal.

File: proposals/0388-bls12-381-syscalls.md (hunk 14 above)
Comment (samkim-crypto) [reply proposals/0388-bls12-381-syscalls.md:249]: I was looking at the design again and yes, I agree that the `num_pairs` design could be slightly cleaner. I updated the function signature accordingly.

**Comment by deanmlittle** [comment]
@0x0ece any objections to approving this?

**Review by 0x0ece — APPROVED** [review]

File: proposals/0388-bls12-381-syscalls.md [hunk 19]
```diff
@@ -0,0 +1,341 @@
...
+1. Field Check: The coordinate(s) are valid field elements (i.e., less than the
+   modulus `p`).
+2. On-Curve Check: The point satisfies the curve equation.
+3. Subgroup Check: The point is in the correct prime-order subgroup `r`.
+
+This syscall is specifically for uncompressed affine points. To validate a
+compressed point, programs can use the `sol_curve_decompress` syscall, which
+performs identical validation checks as part of its decompression process.
+
+While it is possible to validate an uncompressed affine poinnt `(x, y)` by using
```
Comment (Rexicon226) [inline proposals/0388-bls12-381-syscalls.md]:
```suggestion
While it is possible to validate an uncompressed affine point `(x, y)` by using
```

**Review by Rexicon226 — APPROVED** [review]

**Review by jacobcreech — APPROVED** [review]
